src/Thentos/Backend/Api/Adhocracy3Sso.hs

```diff
@@ -175,46 +176,45 @@ githubKey = OAuth2 { oauthClientId = "c4c9355b9ea698f622ba"
                    , oauthAccessTokenEndpoint = "https://github.com/login/oauth/access_token"
                    }
 
+addNewSsoToken :: AC.Action DB SsoToken
+addNewSsoToken = do
+    tok <- A.freshSsoToken
+    AC.update'P $ AddSsoToken tok
+    return tok
 
 -- | FIXME: document!
 githubRequest :: AC.Action DB AuthRequest
 githubRequest = do
-    state <- A.freshRandomName
-
-    -- FIXME: store state in DB.
-
-    return . AuthRequest . cs $ authorizationUrl githubKey `appendQueryParam` [("state", cs state)]
+    state <- addNewSsoToken
+    return . AuthRequest . cs $ authorizationUrl githubKey `appendQueryParam` [("state", cs $ fromSsoToken state)]
 
 
 -- | FIXME: document!
 githubConfirm :: ST -> ST -> AC.Action DB A3.RequestResult
 githubConfirm state code = do
-
-    -- FIXME: lookup state in DB and crash if it does not exist.  remove if it does exist.
-
+    ssoTokenExists <- AC.update'P $ LookupAndRemoveSsoToken (SsoToken state)
```

Comment:
Ok. I put it here because I thought it was too special-purpose to go into `Thentos.Action`, but can move it. Not sure about the LIO stuff, will look into it.